pkg/bff/users.go

```diff
@@ -189,21 +194,23 @@ func (s *Server) Login(c *gin.Context) {
 		collaborator.JoinedAt = collaborator.LastLogin
 	}
 
-	if userRole == TSPRole {
-		// TSP users can be added to multiple organizations
+	if claims.HasPermission(auth.SwitchOrganizations) {
+		// If the user can be in multiple organizations, add the selected organization
+		// to the user's list.
 		appdata.AddOrganization(org.Id)
 	} else {
-		// Organizations with one collaborator need a leader to add other collaborators
-		if userRole == CollaboratorRole && len(org.Collaborators) == 1 {
-			userRole = LeaderRole
+		// Make sure users are able to add collaborators if they are the only member of
+		// their organization.
+		if !claims.HasPermission(auth.UpdateCollaborators) && len(org.Collaborators) == 1 {
+			userRole = auth.LeaderRole
 		}
 
 		// Non-TSP users can only exist in one organization
 		if appdata.OrgID != "" && org.Id != appdata.OrgID {
-			// When switching to a new organization, make sure leader roles are not
-			// unintentionally preserved
-			if userRole == LeaderRole && len(org.Collaborators) > 1 {
-				userRole = CollaboratorRole
+			// Users should not be able to add collaborators if they are migrated to an
+			// existing organization.
+			if claims.HasPermission(auth.UpdateCollaborators) && len(org.Collaborators) > 1 {
+				userRole = auth.CollaboratorRole
```

Comment:
Here is the main part of the PR where we check the user permission in the claims rather than checking the user role.